README.md
# seneca-nats-transport

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url] [![Coverage][coverage-image]][coverage-url]

Seneca NATS transport.

[Seneca](http://senecajs.org/) is a microservices framework and [NATS](http://nats.io/) is 
an open-source, high-performance, lightweight cloud messaging system. This library provides
a publish-subscribe message distribution model.

## Installation

```bash
npm install seneca-nats-transport
```

## Usage

[gnatsd server](http://nats.io/download/) **should** be running.

```javascript
// server.js

require('seneca')()
  .use('nats-transport')
  .add({role: 'foo', cmd: 'bar'}, function(msg, done) { return done(null, msg); })
  .listen({type:'nats'});
```

```javascript
// client.js

require('seneca')()
  .use('nats-transport')
  .client({type:'nats'})
  .act({role: 'foo', cmd: 'bar', arg1: 1, arg2: 2}, console.log);
```

```javascript
// nats.js

require('nats').connect().subscribe('*', function(msg) {
  console.log(msg);
});
```

```bash
gnatsd
node nats.js
node server.js
node client.js
```

## License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/seneca-nats-transport
[npm-image]: https://badge.fury.io/js/seneca-nats-transport.svg

[travis-url]: https://travis-ci.org/devfacet/seneca-nats-transport
[travis-image]: https://travis-ci.org/devfacet/seneca-nats-transport.svg?branch=master

[coverage-url]: https://coveralls.io/github/devfacet/seneca-nats-transport?branch=master
[coverage-image]: https://coveralls.io/repos/github/devfacet/seneca-nats-transport/badge.svg?branch=master


CHANGELOG.md
# Changelog

## 2.0.3 (2017-03-05)

- Upgrade NATS to v0.7.x
- Update travis.yml (gnatsd v0.9.6)
- Maintenance

## 2.0.2 (2016-03-20)

- Maintenance

## 2.0.1 (2016-02-04)

- Misc. updates
- Update devDependencies

## 2.0.0 (2015-12-12)

- Upgrade NATS to v0.5.x
- Support for Seneca v0.8.x
- Remove Node v0.10 support
- Add Coveralls integration
- Upgrade dev dependencies

## 1.1.1 (2015-10-19)

- Add `NATS_SERVERS` and `NATS_URL` env vars

## 1.1.0 (2015-10-19)

- Add servers option for NATS
- Update dev dependencies
- Update package scripts

## 1.0.2 (2015-08-08)

- Remove lodash

## 1.0.1 (2015-08-08)

- Add tests

## 1.0.0 (2015-08-08)

- Initial release

test/test-all.js
/* jslint node: true */
/* global describe: false, it: false, beforeEach: false, afterEach: false */
'use strict';

var lib    = require('../'),
    seneca = require('seneca'),
    nats   = require('nats'),
    expect = require('chai').expect;

// Tests

describe('lib', function() {

  var server,
      client,
      nc;

  beforeEach(function() {
    server = seneca({log:'silent'});
    client = seneca({log:'silent'});
    nc     = nats.connect();
  });

  afterEach(function() {
    client.close();
    server.close();
    nc.close();

    client = null;
    server = null;
    nc     = null;
  });

  it('should listen messages', function(done) {

    var pattern = {role: 'foo', cmd: 'bar'},
        message = {role: 'foo', cmd: 'bar', argNum: 1, argStr: '2', argBool: true, argObj: {}, argAry: []};

    nc.subscribe('seneca_any_act', function(msg) {
      expect(JSON.parse(msg).act).to.deep.equal(message);
      done();
    });

    server
      .use(lib)
      .add(pattern, function(msg, done) { return done(null, msg); })
      .listen({type:'nats'});

    client
      .use(lib)
      .client({type:'nats'})
      .act(message);
  });

  it('should send messages', function(done) {

    var pattern = {role: 'foo', cmd: 'bar'},
        message = {role: 'foo', cmd: 'bar', argNum: 1, argStr: '2', argBool: true, argObj: {}, argAry: []};

    nc.subscribe('seneca_any_res', function(msg) {
      var res = JSON.parse(msg).res;
      expect(res).to.be.a('object');
      expect(res.role).to.equal('foo');
      expect(res.cmd).to.equal('bar');
      expect(res.argNum).to.equal(1);
      expect(res.argStr).to.equal('2');
      expect(res.argBool).to.equal(true);
      expect(res.argObj).to.be.a('object');
      expect(res.argAry).to.be.a('array');
      done();
    });

    server
      .use(lib)
      .add(pattern, function(msg, done) { return done(null, msg); })
      .listen({type:'nats'});

    client
      .use(lib)
      .client({type:'nats'})
      .act(message);
  });

});

test/test-lib.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var lib    = require('../'),
    seneca = require('seneca'),
    nats   = require('nats'),
    expect = require('chai').expect;

// Tests

describe('lib', function() {

  var server,
      client,
      nc;

  beforeEach(function() {
    server = seneca({log:'silent'});
    client = seneca({log:'silent'});
    nc     = nats.connect();
  });

  afterEach(function() {
    client.close();
    server.close();
    nc.close();

    client = null;
    server = null;
    nc     = null;
  });

  it('should listen messages', function(done) {

    var pattern = {role: 'foo', cmd: 'bar'},
        message = {role: 'foo', cmd: 'bar', argNum: 1, argStr: '2', argBool: true, argObj: {}, argAry: []};

    nc.subscribe('seneca_any_act', function(msg) {
      expect(JSON.parse(msg).act).to.deep.equal(message);
      done();
    });

    server
      .use(lib)
      .add(pattern, function(msg, done) { return done(null, msg); })
      .listen({type:'nats'});

    client
      .use(lib)
      .client({type:'nats'})
      .act(message);
  });

  it('should send messages', function(done) {

    var pattern = {role: 'foo', cmd: 'bar'},
        message = {role: 'foo', cmd: 'bar', argNum: 1, argStr: '2', argBool: true, argObj: {}, argAry: []};

    nc.subscribe('seneca_any_res', function(msg) {
      var res = JSON.parse(msg).res;
      expect(res).to.be.a('object');
      expect(res.role).to.equal('foo');
      expect(res.cmd).to.equal('bar');
      expect(res.argNum).to.equal(1);
      expect(res.argStr).to.equal('2');
      expect(res.argBool).to.equal(true);
      expect(res.argObj).to.be.a('object');
      expect(res.argAry).to.be.a('array');
      done();
    });

    server
      .use(lib)
      .add(pattern, function(msg, done) { return done(null, msg); })
      .listen({type:'nats'});

    client
      .use(lib)
      .client({type:'nats'})
      .act(message);
  });

});

index.js
/*
 * seneca-nats-transport
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var nats = require('nats');

module.exports = function(options) {

  var NATS_SERVERS = process.env.NATS_SERVERS,
      NATS_URL     = process.env.NATS_URL;

  var seneca = this,
      plugin = 'nats-transport';

  var senecaOpts  = seneca.options(),
      transpUtils = seneca.export('transport/utils');

  options = seneca.util.deepextend({
    nats: {
      reconnect:            true,
      maxReconnectAttempts: 9999,
      reconnectTimeWait:    1000
    }
  }, senecaOpts.transport, options);

  // Check nats servers
  if(!options.nats.servers && NATS_SERVERS) {
    options.nats.servers = NATS_SERVERS.split(',');
  }

  if(options.nats.servers) {
    if(options.nats.servers instanceof Array) {
      for(var i = 0, len = options.nats.servers.length; i < len; i++) {
        if(typeof options.nats.servers[i] === 'string' && options.nats.servers[i].indexOf('nats://') !== 0) {
          options.nats.servers[i] = 'nats://' + options.nats.servers[i];
        }
      }
    }
  }

  // Check nats url
  if(!options.nats.url && NATS_URL) {
    options.nats.url = NATS_URL;
  }

  if(options.nats.url) {
    if(typeof options.nats.url === 'string' && options.nats.url.indexOf('nats://') !== 0) {
      options.nats.url = 'nats://' + options.nats.url;
    }
  }

  // Listen hook for the transport
  seneca.add({role: 'transport', type: 'nats', hook: 'listen'}, function(msg, done) {

    var seneca     = this,
        type       = msg.type,
        clientOpts = seneca.util.clean(seneca.util.deepextend({}, options[type], msg)),
        clientName = 'listen-' + type,
        nc         = nats.connect(options[type]);

    // Connect event
    nc.on('connect', function(/*client*/) {
      seneca.log.info('listen', 'open', clientOpts);
    });

    // Error event
    nc.on('error', function(err) {
      seneca.log.error('listen', 'error', err);
    });

    // Listen topics
    transpUtils.listen_topics(seneca, msg, clientOpts, function(topic) {
      var topicAct = topic + '_act',
          topicRes = topic + '_res';

      // Subscribe to act topic
      nc.subscribe(topicAct, function(msg) {
        seneca.log.debug('listen', 'subscribe', topicAct, 'message', msg);

        // Handle request
        transpUtils.handle_request(seneca, transpUtils.parseJSON(seneca, clientName, msg), clientOpts, function(out) {
          // If there is an output then
          if(out) {
            // Publish it to response topic
            nc.publish(topicRes, transpUtils.stringifyJSON(seneca, clientName, out));
          }
        });
      });
      seneca.log.info('listen', 'subscribe', topicAct);
    });

    // Closer action
    seneca.add({role: 'seneca', cmd: 'close'}, function(args, cb) {
      seneca.log.debug('listen', 'close', clientOpts);

      nc.close();
      this.prior(args, cb);
    });

    done();
  });

  // Client hook for the transport
  seneca.add({role: 'transport', type: 'nats', hook: 'client'}, function(msg, done) {

    var seneca     = this,
        type       = msg.type,
        clientOpts = seneca.util.clean(seneca.util.deepextend({}, options[type], msg)),
        clientName = 'client-' + type,
        nc         = nats.connect(options[type]);

    // Connect event
    nc.on('connect', function(/*client*/) {
      seneca.log.info('client', 'open', clientOpts);
    });

    // Error event
    nc.on('error', function(err) {
      seneca.log.error('client', 'error', err);
    });

    // Send is called for per topic
    function send(spec, topic, sendDone) {
      var topicAct = topic + '_act',
          topicRes = topic + '_res';

      // Subscribe to response topic
      nc.subscribe(topicRes, function(msg) {
        seneca.log.debug('client', 'subscribe', topicRes, 'message', msg);

        // Handle response
        transpUtils.handle_response(seneca, transpUtils.parseJSON(seneca, clientName, msg), clientOpts);
      });
      seneca.log.info('client', 'subscribe', topicRes);

      // Send message over the transport
      sendDone(null, function(msg, cb) {
        seneca.log.debug('client', 'publish', topicAct, 'message', msg);

        // Publish act
        nc.publish(topicAct, transpUtils.stringifyJSON(seneca, clientName, transpUtils.prepare_request(seneca, msg, cb)));
      });

      // Closer action
      seneca.add({role: 'seneca', cmd: 'close'}, function(args, cb) {
        seneca.log.debug('client', 'close', clientOpts, 'topic', topic);

        nc.close();
        this.prior(args, cb);
      });
    }

    // Use transport utils to make client
    transpUtils.make_client(send, clientOpts, done);
  });

  // Return
  return {
    name: plugin
  };
};

lib/nats-transport.js
/*
 * seneca-nats-transport
 * Copyright (c) 2015 Fatih Cetinkaya (http://github.com/fatihcode/seneca-nats-transport)
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var nats = require('nats');

module.exports = function(options) {

  var NATS_SERVERS = process.env.NATS_SERVERS,
      NATS_URL     = process.env.NATS_URL;

  var seneca = this,
      plugin = 'nats-transport';

  var senecaOpts  = seneca.options(),
      transpUtils = seneca.export('transport/utils');

  options = seneca.util.deepextend({
    nats: {
      reconnect:            true,
      maxReconnectAttempts: 9999,
      reconnectTimeWait:    1000
    }
  }, senecaOpts.transport, options);

  // Check nats servers
  if(!options.nats.servers && NATS_SERVERS) {
    options.nats.servers = NATS_SERVERS.split(',');
  }

  if(options.nats.servers) {
    if(options.nats.servers instanceof Array) {
      for(var i = 0, len = options.nats.servers.length; i < len; i++) {
        if(typeof options.nats.servers[i] === 'string' && options.nats.servers[i].indexOf('nats://') !== 0) {
          options.nats.servers[i] = 'nats://' + options.nats.servers[i];
        }
      }
    }
  }

  // Check nats url
  if(!options.nats.url && NATS_URL) {
    options.nats.url = NATS_URL;
  }

  if(options.nats.url) {
    if(typeof options.nats.url === 'string' && options.nats.url.indexOf('nats://') !== 0) {
      options.nats.url = 'nats://' + options.nats.url;
    }
  }

  // Listen hook for the transport
  seneca.add({role: 'transport', type: 'nats', hook: 'listen'}, function(msg, done) {

    var seneca     = this,
        type       = msg.type,
        clientOpts = seneca.util.clean(seneca.util.deepextend({}, options[type], msg)),
        clientName = 'listen-' + type,
        nc         = nats.connect(options[type]);

    // Connect event
    nc.on('connect', function(/*client*/) {
      seneca.log.info('listen', 'open', clientOpts);
    });

    // Error event
    nc.on('error', function(err) {
      seneca.log.error('listen', 'error', err);
    });

    // Listen topics
    transpUtils.listen_topics(seneca, msg, clientOpts, function(topic) {
      var topicAct = topic + '_act',
          topicRes = topic + '_res';

      // Subscribe to act topic
      nc.subscribe(topicAct, function(msg) {
        seneca.log.debug('listen', 'subscribe', topicAct, 'message', msg);

        // Handle request
        transpUtils.handle_request(seneca, transpUtils.parseJSON(seneca, clientName, msg), clientOpts, function(out) {
          // If there is an output then
          if(out) {
            // Publish it to response topic
            nc.publish(topicRes, transpUtils.stringifyJSON(seneca, clientName, out));
          }
        });
      });
      seneca.log.info('listen', 'subscribe', topicAct);
    });

    // Closer action
    seneca.add({role: 'seneca', cmd: 'close'}, function(args, cb) {
      seneca.log.debug('listen', 'close', clientOpts);

      nc.close();
      this.prior(args, cb);
    });

    done();
  });

  // Client hook for the transport
  seneca.add({role: 'transport', type: 'nats', hook: 'client'}, function(msg, done) {

    var seneca     = this,
        type       = msg.type,
        clientOpts = seneca.util.clean(seneca.util.deepextend({}, options[type], msg)),
        clientName = 'client-' + type,
        nc         = nats.connect(options[type]);

    // Connect event
    nc.on('connect', function(/*client*/) {
      seneca.log.info('client', 'open', clientOpts);
    });

    // Error event
    nc.on('error', function(err) {
      seneca.log.error('client', 'error', err);
    });

    // Send is called for per topic
    function send(spec, topic, sendDone) {
      var topicAct = topic + '_act',
          topicRes = topic + '_res';

      // Subscribe to response topic
      nc.subscribe(topicRes, function(msg) {
        seneca.log.debug('client', 'subscribe', topicRes, 'message', msg);

        // Handle response
        transpUtils.handle_response(seneca, transpUtils.parseJSON(seneca, clientName, msg), clientOpts);
      });
      seneca.log.info('client', 'subscribe', topicRes);

      // Send message over the transport
      sendDone(null, function(msg, cb) {
        seneca.log.debug('client', 'publish', topicAct, 'message', msg);

        // Publish act
        nc.publish(topicAct, transpUtils.stringifyJSON(seneca, clientName, transpUtils.prepare_request(seneca, msg, cb)));
      });

      // Closer action
      seneca.add({role: 'seneca', cmd: 'close'}, function(args, cb) {
        seneca.log.debug('client', 'close', clientOpts, 'topic', topic);

        nc.close();
        this.prior(args, cb);
      });
    }

    // Use transport utils to make client
    transpUtils.make_client(send, clientOpts, done);
  });

  // Return
  return {
    name: plugin
  };

};